packages/office-ui-fabric-react/src/components/DetailsList/DetailsList.tsx

```diff
@@ -458,7 +458,7 @@ export class DetailsList extends BaseComponent<IDetailsListProps, IDetailsListSt
     let { onRowDidMount } = this.props;
     let index = row.props.itemIndex;
 
-    this._activeRows[index] = row; // this is used for column auto resize
+    this._activeRows.splice(index, 0, row);
```

Comment:
why is this better than a map?

Reply:
With the map, if a new row is added at index 0 it overwrites the existing row. 
In addition, if you have a map with keys from 0 to 3, you delete the item with key 2, the key of the next item remains 3. Thus you end up with keys 0,1,3. This implementation maintains _activeRows correctly when adding and removing rows.

We are using this for Rules functionality in VSTS TFS and when we edit the list, the _activeRows was not being maintained correctly and resulting in unexpected functionality.

Reply:
@cschleiden @ThomasMichon Are you guys ok with this change? I ...think... it is ok, but i'm not sure why it's necessary or if it actually hurts perf, without thoroughly debugging through this.

Reply:
I briefly chatted with Chris, but I haven't fully understood the problem yet, asked for repro in the issue https://github.com/OfficeDev/office-ui-fabric-react/issues/2676

Reply:
Replied to the issue, but figured I'd post my response here as well

> When items changes, we do not do a full re-render. All rows mount only when the page is initially loaded. If a row is added or deleted, _onRowDidMount or _onRowWillUnmount is called only for the added or removed row, respectively. This means that _activeRows is not maintained correctly and the keys do not correlate directly to indices. This results in discontinuity in keys when a row is delete and overwriting the initial row when a new row is added. If you want to see this effect, it's visible in the current Rules for Work Item Types implementation in VSTS. Try adding or deleting a rule and see if _activeRows gets updated. I am unsure how to use codepen or what you would be looking to actually have uploaded. If you have an example for my to mirror that would be great and I'd be happy to do that.

Reply:
@dzearing what exactly are your performance concerns? That doing this many splices will be unperformant? 

@cschleiden to understand the issue, imagine you had a grid with 10 items. Then we handed the grid a new list with 11 items, with a new item at the front. Currently, the list will l mount a new row at index 0, overwrite key 0 of the map _activeRows, which will now only have 10 entries even though there are 11 rows in the grid. The same is the case for delete, right now it literally says "delete _activeRows[index]", and doesn't reshift the _activeRows map. So now all of a sudden if I give a new set key and say "focus on index 4 of my list of 5 items", index 4 might be a different entry in _activeRows than the consumer would expect.

If you want to see for yourself, try adding a new rule on the rules implementation of details list here and see how when you add a new row to index 0 how _activeRows gets out of sync: https://mseng.visualstudio.com/DefaultCollection/_admin/_process?process-name=Agile%20VSCS%20Recruiting%20-%20Contractors&type-id=AgileVSCSRecruiting.Candidate&_a=rules

Reply:
@cschleiden as far as i can tell we do not re render all rows when props.items change if setkey does not change, maybe that is the real bug?

Reply:
What happens when you add rows in the middle, will that all work?

Reply:
@cschleiden well in the current implementation if you added a row in the middle, this would just set_activeRows at whatever index you had just added, overriding the row that used to be at that index without ever adding a new entry to _activeRows. It's the same issue when you add to the front. The only time the current implementation works is if you only add to the end. 

With the proposed changes, we would splice the new row into the index that it is currently at, and the _activeRows array will now correctly match props.items.

Reply:
Keys might be more stable (if provided), tried it here https://github.com/OfficeDev/office-ui-fabric-react/pull/2778